pkg/rekor/v2/verifier.go

```diff
@@ -0,0 +1,180 @@
+	// - Store the last checkpoint of the *newest* shard as the last-seen checkpoint
+	// in order to migrate from the old shald to the new one.
+
+	// Fetch (and verify) the latest checkpoint of the latest active shard
+	// This is the checkpoint that will be saved to `logInfoFile`.
+	latestActiveCheckpoint, _, err := (*rekorShards[activeShardOrigin].client).ReadCheckpoint(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get current checkpoint: %v", err)
+	}
+
+	var prevCheckpoint *log.Checkpoint
+	fi, err := os.Stat(logInfoFile)
+	if err == nil && fi.Size() != 0 {
+		// Read the latest saved checkpoint from the log file
+		prevCheckpoint, err = file.ReadLatestCheckpointRekorV2(logInfoFile)
+		if err != nil {
+			return nil, fmt.Errorf("reading checkpoint log: %v", err)
+		}
+
+		// The new checkpoint we fetch for the consistency check has to be from the same
+		// shard as the previous checkpoint.
+		rekorClient := *rekorShards[activeShardOrigin].client
+		newCheckpoint := latestActiveCheckpoint
+		if prevCheckpoint.Origin != activeShardOrigin {
+			rekorClient = *rekorShards[prevCheckpoint.Origin].client
+			newCheckpoint, _, err = rekorClient.ReadCheckpoint(ctx)
+			if err != nil {
+				return nil, fmt.Errorf("failed to get current checkpoint: %v", err)
+			}
+		}
+
+		// Build the consistency proof between the tree sizes of the previous (stored)
+		// checkpoint and the newest fetched checkpoint
+		pb, err := tclient.NewProofBuilder(ctx, newCheckpoint.Size, rekorClient.ReadTile)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get proof builder: %v", err)
+		}
+		consistencyProof, err := pb.ConsistencyProof(ctx, prevCheckpoint.Size, newCheckpoint.Size)
+		if err != nil {
+			return nil, fmt.Errorf("failed to build consistency proof: %v", err)
+		}
+
+		err = proof.VerifyConsistency(rfc6962.DefaultHasher, prevCheckpoint.Size, newCheckpoint.Size, consistencyProof, prevCheckpoint.Hash, newCheckpoint.Hash)
+		if err != nil {
+			return nil, fmt.Errorf("consistency check failed: %v", err)
+		}
+
+		fmt.Fprintf(os.Stderr, "Root hash consistency verified - Current Size: %d Root Hash: %s - Previous Size: %d Root Hash %s\n",
+			newCheckpoint.Size, hex.EncodeToString(newCheckpoint.Hash), prevCheckpoint.Size, hex.EncodeToString(prevCheckpoint.Hash))
+	}
+
+	// Write if there was no stored checkpoint or the origin/sizes differ
+	if prevCheckpoint == nil || prevCheckpoint.Origin != latestActiveCheckpoint.Origin || prevCheckpoint.Size != latestActiveCheckpoint.Size {
+		if err := file.WriteCheckpointRekorV2(latestActiveCheckpoint, logInfoFile); err != nil {
+			// TODO: Once the consistency check and identity search are split into separate tasks, this should hard fail.
+			// Temporarily skipping this to allow this job to succeed, remediating the issue noted here: https://github.com/sigstore/rekor-monitor/issues/271
+			fmt.Fprintf(os.Stderr, "failed to write checkpoint: %v", err)
+		}
+	}
+
```

Comment:
should we have a call to `DeleteOldCheckpoints` similar to v1?

Reply:
not necessary, since we'll switch to only storing the latest checkpoint anyway